docs/enhancements/config-image-versions.md

```diff
@@ -0,0 +1,143 @@
+---
+title: config-image-versions
+authors:
+  - "@lranjbar"
+creation-date: 2021-10-27
+last-updated: 2021-10-29
+---
+
+# Configurable Image Versions
+
+## Summary
+
+When deploying Assisted Installer a ConfigMap is required to do so. In this
+ConfigMap we have the values of various components of Assisted Installer including
+the images for the application. In the code for assisted-service, we define
+several environment variables for these images. We also provide default 
+values for these images which can differ from the desired value at runtime.
+
+For many container based software applications the version of the container ran 
+is determined at runtime. Consider the simple example of `docker run myrepo/myimage` 
+the docker client will resolve this to `hub.docker.com/myrepo/myimage:latest` and 
+then run the image on your system. The expectation is that image ran would be the 
+version you supplied using latest as the default tag.
+
+If we expand this example out to container orchestration technologies, like Openshift,
+the expectation is that the version supplied to Openshift would be the version 
+that is ran for all images and containers. The expectation is that if the version
+you supplied to Openshift doesn't exist then it would fail to pull the image and 
+throw the `ImagePullBackOff` error.
+
+## Motivation
+
+### Goals
+
+The main goal of this proposal is to have the ability to control the images used in 
+an Assisted Installer deployment using standard container deployment methods.
+
+### Non-Goals
+
+There are default versions of Openshift Container Platform and RHCOS images
+supplied to Assisted Installer. These are not the focus of this discussion only
+how we decide which images to deploy and interact with.
+
+## Proposal
+
+First, if we need to populate these environment variables in the code then 
+Assisted Installer should populate them with the environment variables given 
+in the ConfigMap at runtime.
+
+Second, have errors thrown when the environment variables for the images are 
+blank and not provided in the ConfigMap. If a user doesn't supply a value for a 
+required image we should give a runtime error for the user.
+
+Third, throw warnings (or errors) when the version in the environment variable 
```

Comment:
Maybe the confusion is that mostly you have to define these environment variables at runtime for things to work properly. Right now these environment variables have default values though in the code, so if you don't define them at runtime it falls back to those values. What I'm trying to say is we need to error if those environment variables are null or empty. These values for the environment variables should be defined by the user in the artifacts needed for deployment. 

These values of the environment variables could change at runtime from our perspective as the Assisted Installer development team. Another way of wording this is we can't assume in our code which version of Assisted Installer a user will choose to deploy and run.

As a result these values actually do fall in the ConfigMaps and other deployment artifacts. This is kind of what Igal was getting at in the meeting. These things are defined in many places, having a default value in the code kind of puts it in "one" place if you want to use that version. 

The problem is when you want to change the version of the images you are using you have to define it all those different places anyway. The default values technically only hurt you in that case, because it is very unclear what is being deployed. The deployment itself doesn't throw an error in Openshift for you to debug. In fact, Assisted Installer didn't error at all when the deployment method wasn't set up correctly. It just ran the incorrect version of Assisted Installer.

Reply:
I hope it clarifies.. but maybe it doesn't.

Reply:
Your comment makes sense. But it seems like what you're describing is covered by the paragraph above: "Second, have errors...". It's not clear to me what this third scenario is that's different from the second.

Reply:
Need to think on how I can reword it to be more clear. In essence the default values for these things would still exist, just in the deployment artifacts not in the code itself.

Reply:
I think it's helpful to conceptually separate the software from any runtime configuration, regardless of where that config comes from. I would focus this proposal on the software, and just say we're removing defaults from the software.

As a next layer, various packaging and config management tools can be used to manage the runtime configuration. Those tools might or might not offer defaults with an option for user-supplied config. But that's all in the next layer up, which I don't think this proposal needs to address.